File: September 2026/2026-09-04/93/93-restore-ip-add.py
# 93-restore-ip-add.cpp;help;string-backtracking;leetcopde

class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        result = []

        def backtrack(index, parts):
            # If we have 4 parts
            if len(parts) == 4:
                # All digits must be used
                if index == len(s):
                    result.append(".".join(parts))
                return

            # Try taking 1, 2, or 3 digits
            for length in range(1, 4):
                if index + length > len(s):
                    break

                part = s[index:index + length]

                # No leading zeros
                if len(part) > 1 and part[0] == '0':
                    break

                # Value must be <= 255
                if int(part) > 255:
                    continue

                parts.append(part)
                backtrack(index + length, parts)
                parts.pop()

        backtrack(0, [])
        return result

File: September/2026-09-04/93/93-restore-ip-add.cpp
// 93-restore-ip-add.cpp;help;string-backtracking;leetcopde

class Solution {
public:
    vector<string> result;

    void backtrack(string& s, int index, vector<string>& parts) {
        if (parts.size() == 4) {
            if (index == s.length()) {
                result.push_back(
                    parts[0] + "." + parts[1] + "." +
                    parts[2] + "." + parts[3]
                );
            }
            return;
        }
        for (int len = 1; len <= 3; len++) {
            if (index + len > s.length())
                break;

            string part = s.substr(index, len);

            if (part.length() > 1 && part[0] == '0')
                break;

            if (stoi(part) > 255)
                continue;

            parts.push_back(part);

            backtrack(s, index + len, parts);

            parts.pop_back();
        }
    }

    vector<string> restoreIpAddresses(string s) {
        result.clear();

        vector<string> parts;
        backtrack(s, 0, parts);

        return result;
    }
};